migration-engine/migration-engine-tests/tests/migrations/mysql.rs

```diff
@@ -308,3 +308,54 @@ fn default_current_timestamp_precision_follows_column_precision(api: TestApi) {
         .send_sync()
         .assert_migration("01init", |migration| migration.assert_contents(expected_migration));
 }
+
+#[test_connector(tags(Mysql))]
+fn mysql_apply_migrations_errors_gives_the_failed_sql(api: TestApi) {
+    let dm = "";
+    let migrations_directory = api.create_migrations_directory();
+
+    let migration = r#"
+        CREATE TABLE `Cat` ( id INTEGER PRIMARY KEY );
+
+        DROP TABLE `Emu`;
+
+        CREATE TABLE `Emu` (
+            size INTEGER
+        );
+    "#;
+
+    let migration_name = api
+        .create_migration("01init", dm, &migrations_directory)
+        .draft(true)
+        .send_sync()
+        .modify_migration(|contents| {
+            contents.clear();
+            contents.push_str(migration);
+        })
+        .into_output()
+        .generated_migration_name
+        .unwrap();
+
+    let err = api
+        .apply_migrations(&migrations_directory)
+        .send_unwrap_err()
+        .to_string()
+        .replace(&migration_name, "<migration-name>");
+
+    let expectation = expect![[r#"
+        A migration failed to apply. New migrations can not be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve
+
+        Migration name: <migration-name>
+
+        Database error code: 1051
+
+        Database error:
+        Unknown table 'mysql_apply_migrations_errors_gives_the_failed_sql.Emu'
```

Comment:
👍 

(Although this again is a great example for a terrible MySQL error message... how about just mentioning the command that failed, MySQL :( )

Reply:
We cannot reliably! I tried, I had this version actually running, but it's very problematic, because MySQL doesn't give you the failed row like PostgreSQL does. What I tried first, with the following SQL script:

```sql
CREATE TABLE A (id INT PRIMARY KEY);
CREATE ASDASDSAD;
CREATE TABLE B (id INT PRIMARY KEY);
```

So this is how a migration usually gets sent to the apply function. We execute this in one go, and we can iterate the responses per query, meaning:

- First query returns `Ok`
- Second query returns `Err`
- We do not get anything after the error, so we investigate.

We now know the index of the failed query, so what I did is I made a simple split with `;`, and rendered the failed query `CREATE ASDASDSAD;` with corresponding line numbers. How cool is that! Example of the error back then:

```
A migration failed to apply. New migrations can not be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: <migration-name>

Database error code: 1051

Database error:
Unknown table 'mysql_apply_migrations_errors_gives_the_failed_sql.Emu'

Position:
2   CREATE ASDASDSAD;
```

But, this can break horribly, leading us showing something to the users that is just wrong and confusing. Let's imagine a good example where this breaks:

```sql
CREATE TABLE A (id INT PRIMARY KEY, defaultCode VARCHAR(255) DEFAULT 'fn default_function();');
CREATE ASDASDSAD;
CREATE TABLE B (id INT PRIMARY KEY);
```

Now our split with `;` messes up the index of queries listed. We know that the second query fails, but that pesky default value will just break the whole thing. Now we tell the user that the query that caused the error is `1   ';`.

We can also, of course, detect that we're in a string and just not split with the `;` inside. But then we also need to start looking into escaping rules and sooner rather than later we realize we actually need a real SQL parser to get this done in a way that will not break and cause bugs later in the pipeline.

PostgreSQL's error message just tells you the line number of the faulty SQL, which helps tremendously on displaying it in the error. We do not have that privilege with MySQL, so the best we can do here is to tell the number of the query in the script that fails.

Reply:
Oh, I was talking to MySQL here not you :D 
If they just included the statement in the error, our lives would be soooo much easier.

Reply:
Yeah, and also if we could end all wars, hunger and global warming. Would be great, wouldn't it...